### PR TITLE
Added IAM Role setting support for EKS

### PR DESCRIPTION
**Pull request by ZachNo**
With the new support for resolving remote configuration values, allow setting tagging the service accounts with an AWS IAM Role.

Also added arbitrary env var setting as well

```
helm template gremlin ./gremlin \
>     --namespace gremlin \
>     --set gremlin.hostPID=true \
>     --set gremlin.hostNetwork=true \
>     --set gremlin.collect.processes=true \
>     --set gremlin.container.driver=any \
>     --set gremlin.secret.managed=true \
>     --set gremlin.secret.type=secret \
>     --set gremlin.secret.clusterID="EN5882Test" \
>     --set gremlin.secret.key="arn:aws:ssm:us-west-1:123412341234:parameter/cert" \
>     --set gremlin.secret.teamSecret="arn:aws:ssm:us-west-1:123412341234:parameter/secret" \
>     --set gremlin.serviceAccount.iamRole="arn:aws:iam::123412341234:role/TestRole" \
>     --set gremlin.secret.teamID="arn:aws:ssm:us-west-1:123412341234:parameter/TeamId" \
>     --set "gremlin.env.add=- name: TEST_ENV1
>   value: \"80\"
> - name: TEST_ENV2
>   value: blah" \
>     --set chao.serviceAccount.iamRole="arn:aws:iam::123412341234:role/ChaoRole"

---
# Source: gremlin/templates/chao-service-account.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: chao
  namespace: gremlin
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::123412341234:role/ChaoRole
---
# Source: gremlin/templates/gremlin-service-account.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: gremlin
  namespace: gremlin
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::123412341234:role/TestRole
---
# Source: gremlin/templates/secret.yaml
apiVersion: v1
kind: Secret
metadata:
  name: gremlin-secret
  labels:
    app.kubernetes.io/name: gremlin
    helm.sh/chart: gremlin-0.11.0
    app.kubernetes.io/instance: gremlin
    app.kubernetes.io/managed-by: Helm
    version: v1
type: kubernetes.io/Opaque
stringData:
  GREMLIN_CLUSTER_ID: EN5882Test
  GREMLIN_TEAM_ID: arn:aws:ssm:us-west-1:123412341234:parameter/TeamId
  GREMLIN_TEAM_SECRET: arn:aws:ssm:us-west-1:123412341234:parameter/secret
...
# Source: gremlin/templates/daemonset.yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: gremlin
  namespace: gremlin
  labels:
    app.kubernetes.io/name: gremlin
    helm.sh/chart: gremlin-0.11.0
    app.kubernetes.io/instance: gremlin
    app.kubernetes.io/managed-by: Helm
    version: v1
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: gremlin
  updateStrategy:
    rollingUpdate:
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      labels:
        app.kubernetes.io/name: gremlin
        helm.sh/chart: gremlin-0.11.0
        app.kubernetes.io/instance: gremlin
        app.kubernetes.io/managed-by: Helm
        version: v1
    spec:
      serviceAccountName: gremlin
      hostPID: true
      hostNetwork: true
      containers:
      - name: gremlin
        image: gremlin/gremlin:latest
        args: [ "daemon" ]
        imagePullPolicy: Always
        securityContext:
          capabilities:
            add:
              - KILL
              - NET_ADMIN
              - SYS_BOOT
              - SYS_TIME
              - DAC_READ_SEARCH
              - SYS_ADMIN
              - SYS_PTRACE
              - SETFCAP
              - AUDIT_WRITE
              - MKNOD
              - SYS_CHROOT
              - NET_RAW
        env:
          - name: GREMLIN_TEAM_ID
            valueFrom:
              secretKeyRef:
                name:  gremlin-secret
                key: GREMLIN_TEAM_ID
          - name: GREMLIN_TEAM_SECRET
            valueFrom:
              secretKeyRef:
                name: gremlin-secret
                key: GREMLIN_TEAM_SECRET
          - name: GREMLIN_IDENTIFIER
            valueFrom:
              fieldRef:
                fieldPath: spec.nodeName
          - name: GREMLIN_CLIENT_TAGS
            value:
          - name: GREMLIN_DOCKER_IMAGE
            value: gremlin/gremlin:latest
          - name: GREMLIN_COLLECT_PROCESSES
            value: "true"
          - name: GREMLIN_SERVICE_URL
            value: https://api.gremlin.com/v1
          - name: TEST_ENV1
            value: "80"
          - name: TEST_ENV2
            value: blah
...
```

```
helm template gremlin ./gremlin-integrations \
>     --namespace gremlin \
>     --set gremlin.hostPID=true \
>     --set gremlin.hostNetwork=true \
>     --set gremlin.collect.processes=true \
>     --set gremlin.container.driver=any \
>     --set gremlin.secret.managed=true \
>     --set gremlin.secret.type=secret \
>     --set gremlin.secret.clusterID="EN5882Test" \
>     --set gremlin.secret.key="arn:aws:ssm:us-west-1:123412341234:parameter/cert" \
>     --set gremlin.secret.teamSecret="arn:aws:ssm:us-west-1:123412341234:parameter/secret" \
>     --set serviceAccount.iamRole="arn:aws:iam::123412341234:role/TestRole" \
>     --set gremlin.secret.teamID="arn:aws:ssm:us-west-1:123412341234:parameter/TeamId" \
>     --set "gremlin.env.add=- name: TEST_ENV1
>   value: \"80\"
> - name: TEST_ENV2
>   value: blah"
---
# Source: gremlin-integrations/templates/serviceaccount.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: gremlin-integrations
  namespace: gremlin
  labels:
    helm.sh/chart: gremlin-integrations-0.1.0
    app.kubernetes.io/name: gremlin-integrations
    app.kubernetes.io/instance: gremlin
    app.kubernetes.io/version: "0.0.2"
    app.kubernetes.io/managed-by: Helm
  annotations:

    eks.amazonaws.com/role-arn: arn:aws:iam::123412341234:role/TestRole
---
# Source: gremlin-integrations/templates/secret.yaml
apiVersion: v1
kind: Secret
metadata:
  name: gremlin-integrations-secret
  namespace: gremlin
  labels:
    app.kubernetes.io/name: gremlin-integrations
    helm.sh/chart: gremlin-integrations-0.1.0
    app.kubernetes.io/instance: gremlin
    app.kubernetes.io/managed-by: Helm
    version: v1
type: kubernetes.io/Opaque
stringData:
  GREMLIN_TEAM_ID: arn:aws:ssm:us-west-1:123412341234:parameter/TeamId
  GREMLIN_TEAM_SECRET: arn:aws:ssm:us-west-1:123412341234:parameter/secret
---
# Source: gremlin-integrations/templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: gremlin-gremlin-integrations
  namespace: gremlin
  labels:
    helm.sh/chart: gremlin-integrations-0.1.0
    app.kubernetes.io/name: gremlin-integrations
    app.kubernetes.io/instance: gremlin
    app.kubernetes.io/version: "0.0.2"
    app.kubernetes.io/managed-by: Helm
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: gremlin-integrations
      app.kubernetes.io/instance: gremlin
  template:
    metadata:
      labels:
        app.kubernetes.io/name: gremlin-integrations
        app.kubernetes.io/instance: gremlin
    spec:
      serviceAccountName: gremlin-integrations
      securityContext:
        {}
      containers:
        - name: gremlin-integrations
          securityContext:
            {}
          image: "gremlin/gremlin-integrations:latest"
          imagePullPolicy: Always
          env:
            - name: GREMLIN_TEAM_ID
              valueFrom:
                secretKeyRef:
                  name: gremlin-integrations-secret
                  key: GREMLIN_TEAM_ID
            - name: GREMLIN_SERVICE_URL
              value: https://api.gremlin.com/v1
            - name: GREMLIN_TEAM_SECRET
              valueFrom:
                secretKeyRef:
                  name: gremlin-integrations-secret
                  key: GREMLIN_TEAM_SECRET
            - name: TEST_ENV1
              value: "80"
            - name: TEST_ENV2
              value: blah
          volumeMounts:
      volumes:
```

### CODE REVIEW

File: gremlin/templates/gremlin-service-account.yaml [hunk 1]
```diff
@@ -5,6 +5,10 @@ kind: ServiceAccount
 metadata:
   name: gremlin
   namespace: {{ .Release.Namespace }}
+{{ if .Values.gremlin.serviceAccount.iamRole }}
+  annotations:
+    eks.amazonaws.com/role-arn: {{ .Values.gremlin.serviceAccount.iamRole }}
+{{ end }}
```
Comment (thefirstofthe300) [inline gremlin/templates/gremlin-service-account.yaml:11]: By doing this, we're locking users into the `eks.amazonaws.com/role-arn` annotation being the only annotation supported. I'd prefer we go with the https://github.com/gremlin/helm/pull/88/files#diff-ee3b9fdbccfe887c3de3569decd527ad5f930f6f9c76b2c1086fd11f113ed9b3R9-R12 paradigm of allowing them to set any annotations they want.

1) It opens the door for other potential use cases of annotating service accounts
2) It means we won't have to make chart changes if we choose to support Google Cloud KMS as well.

File: gremlin/values.yaml [hunk 2]
```diff
@@ -114,6 +114,8 @@ gremlin:
     #   Gremlin's requirements (see gremlin.podSecurity.podSecurityPolicy and
     #   gremlin.podSecurity.securityContextConstraints)
     create: true
+    # Cloud IAM role to use to resolve any remote secrets
+    iamRole:
```
Comment (philgebhardt) [inline gremlin/values.yaml:118]: Perhaps something like the following to allow adding any arbitrary environment vars, such as
```yaml
env:
  add:
    - name: IAM_ROLE_ARN
      value: arn:aws-iso:iam::208767367117:role/gremlin-role
```

```suggestion
    iamRole:
  env:
    # gremlin.env.add - 
    # Specify any arbitrary enironment variables to pass to the Gremlin Agent daemonset.
    add:
```

**Review by philgebhardt — COMMENTED** [review]
Left a comment about passing IAM_ROLE_ARN. Similar situation to [this discussion](https://github.com/gremlin/helm/pull/88/files#r1423250261), I suppose we'll want to accept input like that in a flexible way.

File: gremlin/values.yaml (hunk 2 above)
Comment (philgebhardt) [inline gremlin/values.yaml:118]: Perhaps something like the following to allow adding any arbitrary environment vars, such as
```yaml
env:
  add:
    - name: IAM_ROLE_ARN
      value: arn:aws-iso:iam::208767367117:role/gremlin-role
```

```suggestion
    iamRole:
  env:
    # gremlin.env.add - 
    # Specify any arbitrary enironment variables to pass to the Gremlin Agent daemonset.
    add:
```

File: gremlin/templates/chao-service-account.yaml [hunk 3]
```diff
@@ -4,6 +4,10 @@ kind: ServiceAccount
 metadata:
   name: chao
   namespace: {{ .Release.Namespace }}
+{{ if .Values.gremlin.serviceAccount.iamRole }}
```
Comment (philgebhardt) [inline gremlin/templates/chao-service-account.yaml:7]: unlike `gremlin-integrations`, `gremlin` and `chao` install together in the same chart. However, it looks like we currently use a `.Values.gremlin.serviceAccount`, but don't use a corresponding `.Values.chao.serviceAccount`. Now might be the point at which we should do that, so we can facilitate the two serviceAccounts using different IAM roles.

The only other value within this `.Values.gremlin.serviceAccount` object today is `.Values.gremlin.serviceAccount.create`. This file uses `.Values.gremlin.installK8sClient`, which we can change to `.Values.chao.serviceAccount.create`

File: gremlin/values.yaml (hunk 2 above)
Comment (ZachNo) [reply gremlin/values.yaml:113]: Added

File: gremlin/templates/chao-service-account.yaml (hunk 3 above)
Comment (ZachNo) [reply gremlin/templates/chao-service-account.yaml]: Moved installK8sClient to create, and added separate chao iamRole

File: gremlin/templates/gremlin-service-account.yaml (hunk 1 above)
Comment (ZachNo) [reply gremlin/templates/gremlin-service-account.yaml]: Allowed both setting annotations and iamRole to live in annotations

File: gremlin/values.yaml [hunk 4]
```diff
@@ -292,6 +295,15 @@ chao:
 
   podLabels: {}
 
+  serviceAccount:
+    # Determines if the `chao` deployment should be installed. This deployment launches an agent that enabled Kubernetes
+    #   targeting for Gremlin attacks
+    create: true
+    # Cloud IAM role to use to resolve any remote secrets
+    iamRole:
```
Comment (philgebhardt) [inline gremlin/values.yaml:303]: I think we can remove this now that we're exposing this generically via `annotations`, that wait we're kubernetes-platform agnostic for now.

Comment (ZachNo) [reply gremlin/values.yaml:308]: Yeah makes sense

File: gremlin-integrations/templates/deployment.yaml [hunk 5]
```diff
@@ -75,6 +75,9 @@ spec:
             - name: SSL_CERT_DIR
               value: {{ .Values.ssl.certDir }}
           {{- end }}
+          {{- with .Values.gremlin.env.add }}
```
[the remark this answers is not in the document]
Comment (ZachNo) [reply gremlin-integrations/templates/deployment.yaml]: Changed. Also simplified the insertion of it, since fromYaml | toYaml is redundant...

**Review by thefirstofthe300 — APPROVED** [review]

**Review by fred-gremlin — APPROVED** [review]

**Review by philgebhardt — APPROVED** [review]

**Pull request merged**